### PR TITLE
First try to test with modules under OSX with clang v16.0.1

### PR DESCRIPTION
**Pull request by ClausKlein**
I run into problems while linking the module-test:
```
clang++ -g -isysroot /Applications/Xcode.app/Contents/Developer/Platforms/MacOSX.platform/Developer/SDKs/MacOSX13.3.sdk \
  -Wl,-search_paths_first -Wl,-headerpad_max_install_names  test/CMakeFiles/test-module.dir/__/src/fmt.cc.o \
  test/CMakeFiles/module-test.dir/module-test.cc.o test/CMakeFiles/module-test.dir/test-main.cc.o -o bin/module-test \
  test/libtest-main.a  fmt.cc.pcm  test/gtest/libgtest.a

Undefined symbols for architecture x86_64:
  "fmt::v9::vsystem_error(int, fmt::v9::basic_string_view<char>, fmt::v9::basic_format_args<fmt::v9::basic_format_context<fmt::v9::appender, char>>)", referenced from:
      std::__1::system_error fmt::v9::system_error<>(int, fmt::v9::basic_format_string<char>) in libtest-main.a(gtest-extra.cc.o)
  "fmt::v9::file::dup(int)", referenced from:
      output_redirect::output_redirect(__sFILE*) in libtest-main.a(gtest-extra.cc.o)
  "fmt::v9::file::dup2(int)", referenced from:
      output_redirect::output_redirect(__sFILE*) in libtest-main.a(gtest-extra.cc.o)
      output_redirect::restore() in libtest-main.a(gtest-extra.cc.o)
  "fmt::v9::file::pipe(fmt::v9::file&, fmt::v9::file&)", referenced from:
      output_redirect::output_redirect(__sFILE*) in libtest-main.a(gtest-extra.cc.o)
  "fmt::v9::file::read(void*, unsigned long)", referenced from:
      output_redirect::restore_and_read() in libtest-main.a(gtest-extra.cc.o)
      read(fmt::v9::file&, unsigned long) in libtest-main.a(gtest-extra.cc.o)
  "fmt::v9::file::close()", referenced from:
      fmt::v9::file::operator=(fmt::v9::file&&) in libtest-main.a(gtest-extra.cc.o)
      output_redirect::restore() in libtest-main.a(gtest-extra.cc.o)
      output_redirect::restore_and_read() in libtest-main.a(gtest-extra.cc.o)
  "fmt::v9::file::~file()", referenced from:
      output_redirect::output_redirect(__sFILE*) in libtest-main.a(gtest-extra.cc.o)
      output_redirect::~output_redirect() in libtest-main.a(gtest-extra.cc.o)
ld: symbol(s) not found for architecture x86_64
clang-16: error: linker command failed with exit code 1 (use -v to see invocation)
bash-3.2$ 
```

### CODE REVIEW

File: test/CMakeLists.txt [hunk 1]
```diff
@@ -55,56 +57,54 @@ function(add_fmt_test name)
   add_test(NAME ${name} COMMAND ${name})
 endfunction()
 
-if (FMT_MODULE)
-  return ()
-endif ()
```
Comment (ClausKlein) [inline test/CMakeLists.txt:60]: This disables all tests?

File: CMakeLists.txt [hunk 2]
```diff
@@ -84,29 +102,31 @@ option(FMT_TEST "Generate the test target." ${FMT_MASTER_PROJECT})
...
 
-set(FMT_CAN_MODULE OFF)
-if (CMAKE_CXX_STANDARD GREATER 17 AND MSVC AND MSVC_VERSION GREATER 1933)
-  set(FMT_CAN_MODULE ON)
+set(FMT_CAN_MODULE ${FMT_MODULE})
+if (CMAKE_CXX_STANDARD GREATER 17)
+  if (MSVC_VERSION GREATER 1933 OR CMAKE_CXX_COMPILER_ID MATCHES "Clang")
+    set(FMT_CAN_MODULE ON)
+  endif ()
 endif ()
```
Comment (ClausKlein) [inline CMakeLists.txt:107]: This seems more clear to me

File: CMakeLists.txt [hunk 3]
```diff
@@ -25,18 +29,32 @@ function(join result_var)
...
+    set(_BMI ${CMAKE_BINARY_DIR}/fmt.cc.pcm)
+    target_compile_options(${target}
+      PRIVATE -fmodule-output=${_BMI} -x c++-module
+      INTERFACE -fmodule-file=fmt=${_BMI})
+    if (${target} MATCHES "fmt")
+      add_custom_command(
+        TARGET fmt POST_BUILD
+        COMMAND ${CMAKE_COMMAND} -E create_symlink ${PROJECT_BINARY_DIR}/CMakeFiles/fmt.dir/src/fmt.cc.pcm ${_BMI}
+        DEPENDS ${FMT_SOURCES})
+    endif ()
```
Comment (ClausKlein) [inline CMakeLists.txt]: Only a workaround, not clear why the a 2 PCM files created with same content?
See `TODO.txt` for my build steps.

Comment (ClausKlein) [reply CMakeLists.txt]: OK, this is fixed now ;-)

**Comment by ClausKlein** [comment]
With this example `CMakeLists.txt`, `CMake v3.26 `works with `clang++ v16.0.1` and `c++20 modules `
But how may this used for the `fmt library as module`?

```cmake
cmake_minimum_required(VERSION 3.26)

set(CMAKE_EXPORT_COMPILE_COMMANDS YES)

project(clang-tidy-issue LANGUAGES CXX)

set(CMAKE_CXX_STANDARD 20)
set(CMAKE_CXX_EXTENSIONS NO)
set(CMAKE_CXX_STANDARD_REQUIRED YES)

set(CMAKE_CXX_SCAN_FOR_MODULES YES)

# First find clang-tidy, this also allows users to provide hints
find_program(CLANG_TIDY NAMES clang-tidy REQUIRED)

# https://clang.llvm.org/docs/StandardCPlusPlusModules.html#how-to-build-projects-using-modules
# clang++ -std=c++20 -x c++-module foo.cxx --precompile -o build/foo.pcm
# clang++ -std=c++20 main.cxx -fprebuilt-module-path=build build/foo.pcm -o build/Hello
#
# see https://gitlab.kitware.com/cmake/cmake/-/blob/v3.26.3/Help/dev/experimental.rst?ref_type=tags#c-20-module-apis
# and https://gitlab.kitware.com/cmake/cmake/-/blob/v3.26.3/.gitlab/ci/cxx_modules_rules_gcc.cmake?ref_type=tags
#     https://gitlab.kitware.com/cmake/cmake/-/blob/v3.26.3/.gitlab/ci/cxx_modules_rules_clang.cmake?ref_type=tags

set(CMAKE_EXPERIMENTAL_CXX_MODULE_CMAKE_API 2182bf5c-ef0d-489a-91da-49dbc3090d2a)
set(CMake_TEST_CXXModules_UUID "a246741c-d067-4019-a8fb-3d16b0c9d1d3")
# Workaround for C++Modules: Missing module map flags in exported compile commands
# see https://gitlab.kitware.com/cmake/cmake/-/issues/24618
set(CMAKE_CXX_COMPILE_OBJECT "${CMAKE_CXX_COMPILE_OBJECT} @<OBJECT>.modmap")

set(CMAKE_EXPERIMENTAL_CXX_MODULE_DYNDEP 1)
if(CMAKE_CXX_COMPILER_ID STREQUAL "GNU" OR CMAKE_CXX_COMPILER_ID STREQUAL "AppleClang")
  string(CONCAT CMAKE_EXPERIMENTAL_CXX_SCANDEP_SOURCE
                "<CMAKE_CXX_COMPILER> <DEFINES> <INCLUDES> <FLAGS> -E -x c++ <SOURCE>"
                " -MT <DYNDEP_FILE> -MD -MF <DEP_FILE>"
                " -fmodules-ts -fdep-file=<DYNDEP_FILE> -fdep-output=<OBJECT> -fdep-format=trtbd"
                " -o <PREPROCESSED_SOURCE>"
  )
  set(CMAKE_EXPERIMENTAL_CXX_MODULE_MAP_FORMAT "gcc")
  set(CMAKE_EXPERIMENTAL_CXX_MODULE_MAP_FLAG "-fmodules-ts -fmodule-mapper=<MODULE_MAP_FILE> -fdep-format=trtbd -x c++")

  set(CMAKE_CXX_CLANG_TIDY "${CLANG_TIDY};-checks=*,-llvmlibc-*,-fuchsia-*,-cppcoreguidelines-init-variables")
elseif(CMAKE_CXX_COMPILER_ID STREQUAL "Clang")
  set(CMAKE_EXPERIMENTAL_CXX_MODULE_MAP_FORMAT "clang")

  # FIXME: how to control the output path of the pcm file with cmake, so that clang-tidy find them too?
  set(CMAKE_CXX_CLANG_TIDY
    "${CLANG_TIDY};-extra-arg=-fprebuilt-module-path=${CMAKE_CURRENT_BINARY_DIR}/CMakeFiles/foo.dir"
  )
endif()

file(WRITE foo.cppm
     [=[
// Global module fragment where #includes can happen
module;
#include <iostream>

// first thing after the Global module fragment must be a module command
// XXX import std.core;

export module foo;

export class foo {
public:
  foo() = default;
  ~foo() = default;
  void helloworld();
};

void foo::helloworld() { std::cout << "hello world\n"; }
]=]
)

file(WRITE main.cxx
     [=[
import foo;

auto main() -> int
{
  foo myFoo;
  myFoo.helloworld();
  return 0;
}
]=]
)

add_library(foo)
target_sources(
  foo
  PUBLIC FILE_SET
         cxx_modules
         TYPE
         CXX_MODULES
         FILES
         foo.cppm
)
add_executable(hello main.cxx)
target_link_libraries(hello foo)

enable_testing()
add_test(NAME hello COMMAND hello)

install(TARGETS foo ARCHIVE PUBLIC_HEADER
  FILE_SET cxx_modules DESTINATION include
  CXX_MODULES_BMI DESTINATION share/cxx-modules
)
include(cpack)
```

**Comment by ClausKlein** [comment]
see too [How to control the location of the c++20 binary-module-interface output-directory?](https://discourse.cmake.org/t/how-to-control-the-location-of-the-c-20-binary-module-interface-bmi-output-directory/7968)

**Comment by DanielaE** [comment]
Cool! 😎
I will look into this after I have clang on MSYS2 running in CI. Until then I will push as many forced updates as needed. As soon as this is working, please rebase on top of that.
Whatever you change, **it must not kill CI**! And please make your changes as few and as small as possible. The smallest viable changeset is the best one. Improvements can come later.
Even better: try to get it on CI, too - if at all possible.

**Comment by DanielaE** [comment]
> With this example `CMakeLists.txt`, `CMake v3.26 `works with `clang++ v16.0.1` and `c++20 modules ` But how may this used for the `fmt library as module`?

This example is from Victor, right?

**Comment by ClausKlein** [comment]
> > With this example `CMakeLists.txt`, `CMake v3.26 `works with `clang++ v16.0.1` and `c++20 modules ` But how may this used for the `fmt library as module`?
> 
> This example is from Victor, right?

It is based on this https://discourse.cmake.org/t/c-20-modules-api-example/7303

**Pull request closed without merging**